### PR TITLE
Bozin Dmitry Lab1 Var2

### PR DESCRIPTION
**Pull request by Nilemos**

### CODE REVIEW

File: clang/labs/lab1/bozin_dmitry/AlwaysInline.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    return false;
+  }
+
+  void outStatusAttrAlwaysInline(clang::FunctionDecl *func) {
+    llvm::outs() << "function: " << func->getNameAsString() << '\n';
+    llvm::outs() << "attr status (always_inline): "
+                 << (func->hasAttr<clang::AlwaysInlineAttr>() ? "true\n"
+                                                              : "false\n");
+  }
+
```
Comment (m-ly4) [inline clang/labs/lab1/bozin_dmitry/AlwaysInline.cpp:55]:
```suggestion
```

File: clang/test/lab1/bozin_dmitry/tests.cpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
+// RUN: %clang_cc1 -load %llvmshlibdir/BozinAlwaysInlinePlugin%pluginext\
+// RUN: -plugin always_inline\
+// RUN: -plugin-arg-always_inline --help %s 2>&1 | FileCheck %s --check-prefix=CHECK-HELP
+// CHECK-HELP: Adds the always_inline attribute to functions without conditions
```
Comment (m-ly4) [inline clang/test/lab1/bozin_dmitry/tests.cpp:4]: Please test that your plugin changes nothing if `help` option is specified.

File: clang/test/lab1/bozin_dmitry/tests.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+// RUN: %clang_cc1 -load %llvmshlibdir/BozinAlwaysInlinePlugin%pluginext\
+// RUN: -add-plugin always_inline %s\
+// RUN: -ast-dump %s -ast-dump-filter test | FileCheck %s
+
+namespace {
+// CHECK: FunctionDecl {{0[xX][0-9a-fA-F]+ <.+tests\.cpp:([0-9]+:[0-9]|[0-9]+), (line|col):([0-9]+:[0-9]|[0-9]+)> (line|col):([0-9]+:[0-9]|[0-9]+) testSquare 'int \(int\)'}}
+int testSquare(int value) { return value * value; }
+
+// CHECK: FunctionDecl {{0[xX][0-9a-fA-F]+ <.+tests\.cpp:([0-9]+:[0-9]|[0-9]+), (line|col):([0-9]+:[0-9]|[0-9]+)> (line|col):([0-9]+:[0-9]|[0-9]+) testDiff 'int \(int, int\)'}}
```
Comment (m-ly4) [inline clang/test/lab1/bozin_dmitry/tests.cpp:14]: Please also check that `always_inline` attribute is added when it is expected.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: clang/labs/lab1/bozin_dmitry/AlwaysInline.cpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  CreateASTConsumer(clang::CompilerInstance &ci, llvm::StringRef) override {
+    return std::make_unique<AddAttrAlwaysInlineConsumer>(&ci.getASTContext());
+  }
+
+  bool ParseArgs(const clang::CompilerInstance &ci,
+                 const std::vector<std::string> &args) override {
+    for (const auto &arg : args) {
+      if (arg == "--help") {
+        llvm::outs() << "Adds the always_inline attribute to functions without "
+                        "conditions\n";
```
Comment (aobolensk) [inline clang/labs/lab1/bozin_dmitry/AlwaysInline.cpp:85]: Only with conditions?

**Review by aobolensk — CHANGES_REQUESTED** [review]

**Pull request closed without merging**